File: generated_code/GPT4/Python/037_SplayTree.py
class SplayTree:
    class Node:
        __slots__ = ['key', 'left', 'right', 'parent']
        def __init__(self, key, parent=None):
            self.key = key
            self.left = None
            self.right = None
            self.parent = parent

    def __init__(self):
        self.root = None

    def _rotate_left(self, x):
        y = x.right
        if y is None:
            return
        x.right = y.left
        if y.left:
            y.left.parent = x
        y.parent = x.parent
        if x.parent is None:
            self.root = y
        else:
            if x == x.parent.left:
                x.parent.left = y
            else:
                x.parent.right = y
        y.left = x
        x.parent = y

    def _rotate_right(self, x):
        y = x.left
        if y is None:
            return
        x.left = y.right
        if y.right:
            y.right.parent = x
        y.parent = x.parent
        if x.parent is None:
            self.root = y
        else:
            if x == x.parent.left:
                x.parent.left = y
            else:
                x.parent.right = y
        y.right = x
        x.parent = y

    def _splay(self, x):
        if x is None:
            return
        while x.parent:
            p = x.parent
            gp = p.parent
            if gp is None:
                # Zig
                if x == p.left:
                    self._rotate_right(p)
                else:
                    self._rotate_left(p)
            else:
                if x == p.left and p == gp.left:
                    # Zig-Zig
                    self._rotate_right(gp)
                    self._rotate_right(p)
                elif x == p.right and p == gp.right:
                    # Zig-Zig
                    self._rotate_left(gp)
                    self._rotate_left(p)
                elif x == p.right and p == gp.left:
                    # Zig-Zag
                    self._rotate_left(p)
                    self._rotate_right(gp)
                elif x == p.left and p == gp.right:
                    # Zig-Zag
                    self._rotate_right(p)
                    self._rotate_left(gp)

    def insert(self, key):
        if self.root is None:
            self.root = self.Node(key)
            return
        node = self.root
        parent = None
        while node:
            parent = node
            if key < node.key:
                node = node.left
            elif key > node.key:
                node = node.right
            else:
                # Key already exists, splay and return
                self._splay(node)
                return
        new_node = self.Node(key, parent)
        if key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
        self._splay(new_node)

    def search(self, key):
        node = self.root
        last = None
        while node:
            last = node
            if key < node.key:
                node = node.left
            elif key > node.key:
                node = node.right
            else:
                self._splay(node)
                return True
        # Not found, splay last accessed node (parent of where it would be)
        if last:
            self._splay(last)
        return False

    def _subtree_minimum(self, node):
        while node.left:
            node = node.left
        return node

    def delete(self, key):
        if self.root is None:
            return
        found = self.search(key)
        if not found:
            return  # Key not found
        # Now, self.root.key == key
        if self.root.left is None:
            self._replace_root(self.root.right)
        elif self.root.right is None:
            self._replace_root(self.root.left)
        else:
            # Both children exist
            right_subtree = self.root.right
            left_subtree = self.root.left
            left_subtree.parent = None
            # Find max in left subtree, splay it
            max_left = left_subtree
            while max_left.right:
                max_left = max_left.right
            self._splay_node_in_subtree(max_left, left_subtree)
            # Now, max_left is root of left_subtree and has no right child
            max_left.right = right_subtree
            if right_subtree:
                right_subtree.parent = max_left
            self.root = max_left
            max_left.parent = None

    def _replace_root(self, node):
        self.root = node
        if node:
            node.parent = None

    def _splay_node_in_subtree(self, node, subtree_root):
        # Splay 'node' to be root of 'subtree_root'
        while node.parent:
            p = node.parent
            gp = p.parent
            if gp is None or gp == subtree_root.parent:
                # Zig
                if node == p.left:
                    self._rotate_right(p)
                else:
                    self._rotate_left(p)
            else:
                if node == p.left and p == gp.left:
                    self._rotate_right(gp)
                    self._rotate_right(p)
                elif node == p.right and p == gp.right:
                    self._rotate_left(gp)
                    self._rotate_left(p)
                elif node == p.right and p == gp.left:
                    self._rotate_left(p)
                    self._rotate_right(gp)
                elif node == p.left and p == gp.right:
                    self._rotate_right(p)
                    self._rotate_left(gp)

    # Optional: for debugging & testing
    def inorder(self):
        res = []
        def _inorder(node):
            if node:
                _inorder(node.left)
                res.append(node.key)
                _inorder(node.right)
        _inorder(self.root)
        return res

File: generated_code/GPT4/Java/037_SplayTree.java
public class SplayTree {
    private class Node {
        int key;
        Node left, right, parent;

        Node(int key, Node parent) {
            this.key = key;
            this.parent = parent;
        }
    }

    private Node root;

    /**
     * Inserts the key into the splay tree.
     * Duplicate keys are ignored.
     */
    public void insert(int key) {
        if (root == null) {
            root = new Node(key, null);
            return;
        }
        Node curr = root, parent = null;
        while (curr != null) {
            parent = curr;
            if (key < curr.key)
                curr = curr.left;
            else if (key > curr.key)
                curr = curr.right;
            else
                return; // Duplicate, do nothing
        }
        Node newNode = new Node(key, parent);
        if (key < parent.key)
            parent.left = newNode;
        else
            parent.right = newNode;
        splay(newNode);
    }

    /**
     * Deletes the key from the splay tree.
     * If the key does not exist, nothing happens.
     */
    public void delete(int key) {
        Node node = findNode(key);
        if (node == null) return;
        splay(node);

        // Now node is root
        if (node.left == null) {
            replaceRoot(node.right);
        } else if (node.right == null) {
            replaceRoot(node.left);
        } else {
            // Both children exist
            Node maxLeft = subtreeMaximum(node.left);
            splay(maxLeft); // Splay the maximum in left subtree to root's left child
            // maxLeft is now root's left child, and has no right child
            maxLeft.right = node.right;
            if (node.right != null)
                node.right.parent = maxLeft;
            replaceRoot(maxLeft);
        }
        // Help GC
        node.left = node.right = node.parent = null;
    }

    /**
     * Searches for the key, splays it to the root if found.
     * Returns the key if found, else null.
     */
    public Integer search(int key) {
        Node node = findNode(key);
        if (node != null) {
            splay(node);
            return node.key;
        }
        return null;
    }

    // --- Helper methods ---

    // Splay the given node to the root
    private void splay(Node x) {
        if (x == null) return;
        while (x.parent != null) {
            Node p = x.parent;
            Node g = p.parent;
            if (g == null) {
                // Zig step
                if (x == p.left)
                    rightRotate(p);
                else
                    leftRotate(p);
            } else if (x == p.left && p == g.left) {
                // Zig-Zig step (left-left)
                rightRotate(g);
                rightRotate(p);
            } else if (x == p.right && p == g.right) {
                // Zig-Zig step (right-right)
                leftRotate(g);
                leftRotate(p);
            } else if (x == p.right && p == g.left) {
                // Zig-Zag step (left-right)
                leftRotate(p);
                rightRotate(g);
            } else if (x == p.left && p == g.right) {
                // Zig-Zag step (right-left)
                rightRotate(p);
                leftRotate(g);
            }
        }
        root = x;
    }

    // Right rotation at node y
    private void rightRotate(Node y) {
        Node x = y.left;
        if (x == null) return;
        y.left = x.right;
        if (x.right != null) x.right.parent = y;
        x.parent = y.parent;
        if (y.parent == null)
            root = x;
        else if (y == y.parent.left)
            y.parent.left = x;
        else
            y.parent.right = x;
        x.right = y;
        y.parent = x;
    }

    // Left rotation at node x
    private void leftRotate(Node x) {
        Node y = x.right;
        if (y == null) return;
        x.right = y.left;
        if (y.left != null) y.left.parent = x;
        y.parent = x.parent;
        if (x.parent == null)
            root = y;
        else if (x == x.parent.left)
            x.parent.left = y;
        else
            x.parent.right = y;
        y.left = x;
        x.parent = y;
    }

    // Find node with key
    private Node findNode(int key) {
        Node curr = root;
        while (curr != null) {
            if (key < curr.key)
                curr = curr.left;
            else if (key > curr.key)
                curr = curr.right;
            else
                return curr;
        }
        return null;
    }

    // Replace root with node (can be null)
    private void replaceRoot(Node node) {
        root = node;
        if (node != null)
            node.parent = null;
    }

    // Returns the maximum node in the given subtree
    private Node subtreeMaximum(Node node) {
        while (node.right != null)
            node = node.right;
        return node;
    }

    // --- Optional: For debugging or testing only ---
    // public void inOrder() { inOrder(root); System.out.println(); }
    // private void inOrder(Node n) {
    //     if (n == null) return;
    //     inOrder(n.left);
    //     System.out.print(n.key + " ");
    //     inOrder(n.right);
    // }
}